library/std/src/sys/dbg.rs

```diff
@@ -0,0 +1,216 @@
+                core::ptr::null_mut(),
+                &mut info_size,
+                core::ptr::null_mut(),
+                0,
+            )
+        } != 0
+        {
+            return Err(DebuggerPresenceError::DetectionFailed);
+        }
+
+        if info_size < KINFO_PROC_SIZE {
+            return Err(DebuggerPresenceError::DetectionFailed);
+        }
+
+        let mut kinfo_proc = vec![0u8; info_size];
+
+        // SAFETY: No state is shared with other threads. The sysctl call
+        // is safe according to the documentation.
+        if unsafe {
+            sysctl(
+                mib.as_mut_ptr(),
+                mib.len() as u32,
+                kinfo_proc.as_mut_ptr().cast(),
+                &mut info_size,
+                core::ptr::null_mut(),
+                0,
+            )
+        } != 0
+        {
+            return Err(DebuggerPresenceError::DetectionFailed);
+        }
+
+        let mut reader = crate::io::Cursor::new(kinfo_proc);
+        reader
+            .seek(crate::io::SeekFrom::Start(KINFO_PROC_FLAGS_OFFSET))
+            .map_err(|_| DebuggerPresenceError::DetectionFailed)?;
+        reader.read_exact(&mut flags).map_err(|_| DebuggerPresenceError::DetectionFailed)?;
+        // Just in case, not limiting this to the little-endian systems.
+        let flags = i32::from_ne_bytes(flags);
+
+        if flags & P_TRACED != 0 {
+            Ok(DebuggerPresence::Detected)
+        } else {
+            Ok(DebuggerPresence::NotDetected)
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+mod os {
+    use super::{DebuggerPresence, DebuggerPresenceError};
+    use crate::fs::File;
+    use crate::io::{BufRead, BufReader};
+
+    pub(super) fn is_debugger_present() -> Result<DebuggerPresence, DebuggerPresenceError> {
+        let file =
+            File::open("/proc/self/status").map_err(|_| DebuggerPresenceError::NotImplemented)?;
+        let reader = BufReader::new(file);
+
+        for line in reader.lines().flatten() {
+            if line.starts_with("TracerPid:") {
+                return if line.split(':').nth(1).map_or(false, |pid| pid.trim() != "0") {
+                    Ok(DebuggerPresence::Detected)
+                } else {
+                    Ok(DebuggerPresence::NotDetected)
+                };
+            }
+        }
+
+        Err(DebuggerPresenceError::DetectionFailed)
+    }
+}
+
+#[cfg(not(any(
+    target_os = "windows",
+    target_os = "macos",
+    target_os = "freebsd",
+    target_os = "linux"
+)))]
+mod os {
+    use super::{DebuggerPresence, DebuggerPresenceError};
+
+    pub(super) fn is_debugger_present() -> Result<DebuggerPresence, DebuggerPresenceError> {
+        Err(DebuggerPresenceError::NotImplemented)
+    }
+}
+
+/// Detect the debugger presence.
+pub(crate) fn is_debugger_present() -> Result<DebuggerPresence, DebuggerPresenceError> {
+    if cfg!(miri) {
+        return Err(DebuggerPresenceError::NotImplemented);
+    }
+    os::is_debugger_present()
```

Comment:
I prefer to avoid this style for the simple if-else case because it typechecks differently.
```suggestion
    if cfg!(miri) {
        Err(DebuggerPresenceError::NotImplemented)
    } else {
        os::is_debugger_present()
    }
```

Reply:
Subtle, love the explanation! Fixed.